misc/releaser.py

```
@@ -382,14 +377,60 @@ def build_release(version):
     run_git("commit", "-m", commit_msg, "--no-verify")
 
 
-def check_release(version):
+def gen_rst_release_entry(
+    version: Version, release_date: date, issues_per_type: defaultdict[str, list[str]]
+) -> str:
+    new_entry_title = f"Parsec {version} ({release_date.isoformat()})"
+    new_entry = f"\n\n{new_entry_title}\n{len(new_entry_title) * '-'}\n"
+
+    if not issues_per_type:
+        new_entry += "\nNo significant changes.\n"
+    else:
+        for fragment_type, fragment_title in FRAGMENT_TYPES.items():
+            if fragment_type not in issues_per_type:
+                continue
+            new_entry += f"\n{fragment_title}\n{len(fragment_title) * '~'}\n\n"
+            new_entry += "\n".join(issues_per_type[fragment_type])
+            new_entry += "\n"
+    return new_entry
+
+
+def convert_newsfragments_to_rst(newsfragments: list[Path]) -> defaultdict[str, list[str]]:
+    issues_per_type = defaultdict(list)
+    for fragment in newsfragments:
+        issue_id, type, _ = fragment.name.split(".")
+        # Don't add empty fragments. Still needed to be collected as they will be deleted later
+        if type == "empty":
+            continue
+        issue_txt = f"{fragment.read_text(encoding='utf8')} (`#{issue_id} <https://github.com/Scille/parsec-cloud/issues/{issue_id}>`__)\n"
+        wrapped_issue_txt = textwrap.fill(
+            issue_txt, width=80, break_long_words=False, initial_indent="* ", subsequent_indent="  "
+        )
+        issues_per_type[type].append(wrapped_issue_txt)
+    return issues_per_type
+
+
+def split_history_file() -> tuple[str, str]:
+    history_txt = HISTORY_FILE.read_text(encoding="utf8")
+    header_split = ".. towncrier release notes start\n"
+    if header_split in history_txt:
+        header, history_body = history_txt.split(header_split, 1)
+        history_header = header + header_split
+    else:
+        # TODO: Should we panic here if we don't find the split tag ?
```

Yes it's better to raise an exception in this case given it most likely indicate something wrong is going on !

---

done